lib/libzfs_core/libzfs_core.c

```diff
@@ -586,6 +589,131 @@ lzc_get_holds(const char *snapname, nvlist_t **holdsp)
 	return (lzc_ioctl(ZFS_IOC_GET_HOLDS, snapname, NULL, holdsp));
 }
 
+static unsigned int
+max_pipe_buffer(int infd)
+{
+#if __linux__
+	unsigned int max = 1048576; /* fs/pipe.c default */
+	FILE *procf = fopen("/proc/sys/fs/pipe-max-size", "re");
+	if (procf != NULL) {
+		(void) fscanf(procf, "%u", &max);
+		fclose(procf);
+	}
+
+	unsigned int cur = fcntl(infd, F_GETPIPE_SZ);
+	if (cur < max && fcntl(infd, F_SETPIPE_SZ, max) != -1)
+		cur = max;
+	return (cur);
+#else
+	/* FreeBSD automatically resizes */
+	(void) infd;
+	return (BIG_PIPE_SIZE);
+#endif
+}
+
+struct send_worker_ctx {
+	int from;	/* read end of pipe, with send data; closed on exit */
+	int to;		/* original arbitrary output fd; mustn't be a pipe */
+};
+
+static void *
+send_worker(void *arg)
+{
+	struct send_worker_ctx *ctx = arg;
+	unsigned int bufsiz = max_pipe_buffer(ctx->from);
+	ssize_t rd;
+	fprintf(stderr, "send worker :0 %u\n", bufsiz);
+
+#if __linux__
+	while ((rd = splice(ctx->from, NULL, ctx->to, NULL, bufsiz,
+	    SPLICE_F_MOVE | SPLICE_F_MORE)) > 0)
+		;
+#else
+	void *abuf = malloc(bufsiz);
+	void *buf = abuf ?: (bufsiz = 4096, alloca(bufsiz));
+	for (;;) {
+		while ((rd = read(ctx->in, buf, bufsiz)) == -1 &&
+		    errno == EINTR)
+			;
+		if (rd <= 0)
+			break;
+
+		void *obuf = buf;
+		while (rd != 0) {
+			ssize_t wr;
+			while ((wr = write(ctx->in, obuf, rd)) == -1 &&
+			    errno == EINTR)
+				;
+			if (wr == -1) {
+				rd = wr;
+				goto break2;
+			}
+			obuf += wr;
+			rd -= wr;
+		}
+	}
+break2:
+	free(abuf);
+#endif
+
+	int err = (rd == -1) ? errno : 0;
+	close(ctx->from);
+	return ((void *)(uintptr_t)err);
+}
+
+/*
+ * Since Linux 5.10, 4d03e3cc59828c82ee89ea6e27a2f3cdf95aaadf
+ * ("fs: don't allow kernel reads and writes without iter ops"),
+ * ZFS_IOC_SEND* will EINVAL when writing to /dev/null, /dev/zero, &c.
+ *
+ * This wrapper transparently executes func() with a pipe
+ * by spawning a thread to copy from that pipe to the original output
+ * in the background.
+ *
+ * Returns the error from func(), if nonzero,
+ * otherwise the error from the thread.
+ *
+ * No-op if orig_fd is -1, already a pipe, and on not-Linux;
+ * as such, it is safe to wrap/call wrapped functions in a wrapped context.
+ */
+int
+lzc_send_wrapper(int (*func)(int, void *), int orig_fd, void *data)
+{
+#if __linux__
```

Comment:
I'm a little sad about the notion of only doing the separate thread on Linux unless we know it's going to cause performance regressions - mostly because I fear people encountering bugs that only come up on one platform.

Not strongly enough to suggest changing it, without a concrete example of it breaking, just strongly enough to be curious what you think about doing so.